pylint/checkers/len.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2016 Alexander Todorov <atodorov@MrSenko.com>
+
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/master/COPYING
+
+"""check for incorrect usage of len() in conditions
+"""
+import itertools
+from astroid.node_classes import Call, Const, Name, UnaryOp
+
+from pylint.interfaces import IAstroidChecker
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import check_messages
+
+MSGS = {
+    'W1801': ('Do not use `len(SEQUENCE)` as condition value',
+              'len-as-condition',
+              'Used when Pylint detects incorrect use of len(sequence) inside \
+              conditions.'),
+    }
+
+
+def _is_len_call(node):
+    """checks if @node is len(SOMETHING)"""
+    return isinstance(node, Call) and isinstance(node.func, Name) and node.func.name == 'len'
+
+
+class LenChecker(BaseChecker):
+    """checks for incorrect usage of len() inside conditions. pep8 states:
+    For sequences, (strings, lists, tuples), use the fact that empty sequences are false.
+
+        Yes: if not seq:
+             if seq:
+
+        No: if len(seq):
+            if not len(seq):
+
+    Problems detected:
+    * if len(sequence):
+    * if not len(sequence):
+    * if len(sequence) == 0:
+    * if len(sequence) != 0:
+    * if len(sequence) > 0:
+    """
+
+    __implements__ = (IAstroidChecker,)
+
+    # configuration section name
+    name = 'len'
+    # messages
+    msgs = MSGS
+    priority = -2
+    # configuration options
+    options = ()
+
+    @check_messages('len-as-condition')
+    def visit_call(self, node):
+        # a len(S) call is used inside a test condition
+        # could be if, while, assert or if expression statement
+        # e.g. `if len(S):`
+        if hasattr(node.parent, 'test') and _is_len_call(node):
+            self.add_message('len-as-condition', node=node)
+
+    @check_messages('len-as-condition')
+    def visit_unaryop(self, node):
+        # `not len(S)` must become `not S` regardless if the parent block
+        # is a test condition or something else (boolean expression)
+        # e.g. `if not len(S):`
+        if isinstance(node, UnaryOp) and node.op == 'not' and \
+            _is_len_call(node.operand):
+            self.add_message('len-as-condition', node=node)
+
+    @check_messages('len-as-condition')
+    def visit_compare(self, node):
+        # compare nodes are trickier because the len(S) expression
+        # may be somewhere in the middle of the node
+        ops = [('', node.left)]
+        ops.extend(node.ops)
+        ops = list(itertools.chain(*ops))
+
+        for ops_idx in range(len(ops)-2):
+            op_1 = ops[ops_idx]
+            op_2 = ops[ops_idx+1]
+            op_3 = ops[ops_idx+2]
+
+            # 0 ?? len()
+            if isinstance(op_1, Const) and (op_1.value == 0) and \
```

Comment:
Consider pulling this out into a function like `_is_constant_zero` or something to reduce code duplication.